### ISSUE TITLE
Can you share your training loss log for Mistral-ORPO-β (7B)?

### ISSUE THREAD

**Issue by yechenzhi** [xfactlab/orpo#26]
I implemented lora ORPO myself using [Torchtune](https://github.com/pytorch/torchtune). However, I'm unsure if it's correct, and the loss curve appears somewhat irregular. Therefore, I'd like to compare it with yours to see if they align.
my version(not finished training):
loss and policy_nll_loss:
![image](https://github.com/xfactlab/orpo/assets/18010625/55e21e4f-bcc7-4ce2-b5ac-a37f33f2e969)
ORPO loss:
![image](https://github.com/xfactlab/orpo/assets/18010625/64430b81-657b-45d0-a7e0-a638f7ec45b7)

**Comment by jiwooya1000** [comment]
Hello @yechenzhi,
for better reproducibility, I just trained `Mistral-orpo-beta` with [alignment-handbook](https://github.com/huggingface/alignment-handbook).
All you need to do is make a new `.yaml` file like this example of [Zephyr-ORPO-141B](https://github.com/huggingface/alignment-handbook/blob/main/recipes/zephyr-141b-A35b/orpo/config_full.yaml).

The corresponding .yaml file will look like this (you could fill in `hub_model_id` and `output_dir` with your values).
```
# Model arguments
model_name_or_path: mistralai/Mistral-7B-v0.1
model_revision: main
torch_dtype: bfloat16
use_flash_attention_2: true

# Data training arguments
chat_template: "{% for message in messages %}\n{% if message['role'] == 'user' %}\n{{ '<|user|>\n' + message['content'] + eos_token }}\n{% elif message['role'] == 'system' %}\n{{ '<|system|>\n' + message['content'] + eos_token }}\n{% elif message['role'] == 'assistant' %}\n{{ '<|assistant|>\n'  + message['content'] + eos_token }}\n{% endif %}\n{% if loop.last and add_generation_prompt %}\n{{ '<|assistant|>' }}\n{% endif %}\n{% endfor %}"
dataset_mixer:
  argilla/ultrafeedback-binarized-preferences-cleaned: 1.0
dataset_splits:
- train
preprocessing_num_workers: 48

# ORPOTrainer arguments
bf16: true
beta: 0.1
gradient_accumulation_steps: 2
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: true
hub_model_id:   
learning_rate: 5.0e-6
log_level: info
logging_steps: 10
lr_scheduler_type: cosine
max_length: 2048
max_prompt_length: 1792
num_train_epochs: 3 
optim: paged_adamw_8bit
output_dir: 
per_device_train_batch_size: 8
push_to_hub: true
report_to:
- tensorboard
- wandb
save_strategy: "epoch"
seed: 42
hub_private_repo: True
```

And the resulting training log will look like this (MT-Bench was 7.325 with this checkpoint, which aligns with the one reported in the paper):
![image](https://github.com/xfactlab/orpo/assets/20739007/131a665b-3a4b-4c38-932f-9707dcdaae1b)

Hope this could be help you comparing the results🙂

**Comment by qingquansong** [comment]
Hey @jiwooya1000 thank you for providing the repo! I'm trying to follow the instruction to use alignment handbook to reproduce and with same config + 4GPUs (the reason I choose 4 is because I saw you have 
~2850 steps in total (3 epochs) so with the training batch size 8 per gpu and grad accumulation 2 and 60.8k samples, it will  roughly take 950 step per epoch) However, after finish the training, the mt-bench results at my side is `Turn1 - 7.21, Turn 2 - 5.95`. One more change I have is the optimizer as there're some dependency issue of using 8bit optimizer, I switch to use the 32bit original torch AdamW optimizer. The losses and log ads ratio I got is below which seems to be much smaller than the one you shown I the end for each:

<img width="1398" alt="Screenshot 2024-05-23 at 4 26 44 PM" src="https://github.com/xfactlab/orpo/assets/25535036/7ded785f-7d87-4371-b66f-1d79dd06e3e4">
![Screenshot 2024-05-23 at 4 42 48 PM](https://github.com/xfactlab/orpo/assets/25535036/e6d3ac89-67d3-4806-a505-77ed3a134b8a)
![Screenshot 2024-05-23 at 4 50 20 PM](https://github.com/xfactlab/orpo/assets/25535036/c837c5db-9ad6-4440-9b40-5dd261195718)


My config is below. Any further instruction to share to reproduce the model? Thank you very much!

```

# Model arguments
model_name_or_path: mistralai/Mistral-7B-v0.1
model_revision: main
torch_dtype: bfloat16
use_flash_attention_2: true

# Data training arguments
chat_template: "{% for message in messages %}\n{% if message['role'] == 'user' %}\n{{ '<|user|>\n' + message['content'] + eos_token }}\n{% elif message['role'] == 'system' %}\n{{ '<|system|>\n' + message['content'] + eos_token }}\n{% elif message['role'] == 'assistant' %}\n{{ '<|assistant|>\n'  + message['content'] + eos_token }}\n{% endif %}\n{% if loop.last and add_generation_prompt %}\n{{ '<|assistant|>' }}\n{% endif %}\n{% endfor %}"
dataset_mixer:
  argilla/ultrafeedback-binarized-preferences-cleaned: 1.0
dataset_splits:
- train
preprocessing_num_workers: 128

# ORPOTrainer arguments
bf16: true
beta: 0.1
gradient_accumulation_steps: 2
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: true
hub_model_id: mistral-7b-orpo-beta
learning_rate: 5.0e-6
log_level: info
logging_steps: 10
lr_scheduler_type: cosine
max_length: 2048
max_prompt_length: 1792
num_train_epochs: 3 
optim: adamw_torch
output_dir: /alignment_handbook_models/mistral-7b-orpo-beta3
per_device_train_batch_size: 8
push_to_hub: false
report_to:
- tensorboard
save_strategy: "epoch"
seed: 42
hub_private_repo: false
```

**Comment by nlee-208** [comment]
Hey @qingquansong, we recently noticed an issue in trl's `OrpoTrainer` in masking out the padded tokens for NLL loss (the nll loss is too low in the diagram you provided). Could you try the replication after updating the trl version to `0.8.7.dev` or higher as stated [here](https://github.com/huggingface/trl?tab=readme-ov-file#from-source)?

**Comment by qingquansong** [comment]
Hey @nlee-208 thank you for the quick response! The version we're using is `Version: 0.8.2`. Let me try the new version you mentioned. Thanks!

**Comment by nlee-208** [comment]
1) I don't happen to know the exact trl version of the reproduced results above but sure it is before the nll problem was incurred in trl. 

2) If you're trying to replicate the `mistral-orpo-beta` checkpoint, it's the setting in the parentheses with total batch size of 32. But considering the previous comments above not sure if that change will be significant.

**Comment by qingquansong** [comment]
Hey @nlee-208 thanks for the repsonse！For (2), I think you mean 64 right? 8*8gpu * grad accum 1 as this should result in 2850 total steps in 3 epochs as shown in @jiwooya1000 's plot?

**Comment by chtmp223** [comment]
Hi @jiwooya1000, I'm using Orpo on my custom dataset and noticed something odd. The initial accuracy isn't around 0.5, which should be random instead since the model hasn't yet differentiated between chosen and rejected responses. Could you share the reward accuracy curve from your experiments? It would help determine if this trend is normal or not. For reference, here is my current reward accuracy curve:

<img width="1283" alt="image" src="https://github.com/xfactlab/orpo/assets/41503463/6b5e1665-26e7-482e-b200-2540f4de777d">

**Comment by jiwooya1000** [comment]
Hello @chtmp223,
the shared situation can indeed happen under a few conditions:

1. **If your dataset has clear pattern/bias**: For example, if chosen has very chatty style with long responses while the rejected responses are very short and non-friendly style, the model can easily capture the difference.
2. **If the dataset is self-generated**: If your dataset is synthetic dataset generated by the trained model or the variant of trained model, there is a possibility which your trained model can more easily capture the latent preference.

Could you share if your custom dataset has such features?

**Comment by chtmp223** [comment]
Thank you for the quick response! The dataset I'm using wasn't generated by the same model that I'm using for fine-tuning (Mistral-Instruct-v0.2). I also don't believe there's a clear pattern that distinguishes between the chosen and rejected responses. They have roughly the same length and formatting. The tone might be slightly different, but it should be negligible.

**Comment by jiwooya1000** [comment]
Thank you for sharing the details!
One thing I have noticed is that Mistral-Instruct-v0.2 uses the left padding by default, but ORPO (and other alignment methods like DPO) requires the right padding. This could be one reason for the high initial reward.
I pushed the changes to force every model's tokenizer to use the right padding.
Could you try it again by pulling the changes?

**Comment by chtmp223** [comment]
Changing the padding side unfortunately doesn't help (the reward accuracy is always 1 now). Thank you for your help though! I will take a closer look at my data.
